lib/listen/event/loop.rb

```diff
@@ -6,97 +6,86 @@
 module Listen
   module Event
     class Loop
+      include Listen::FSM
+
       class Error < RuntimeError
-        class NotStarted < Error
-        end
+        class ThreadFailedToStart < Error; end
+        class AlreadyStarted < Error; end
       end
 
+      start_state :pre_start
+      state :pre_start
+      state :starting
+      state :started
+      state :stopped
+
       def initialize(config)
         @config = config
         @wait_thread = nil
-        @state = :paused
         @reasons = ::Queue.new
+        super()
       end
 
       def wakeup_on_event
-        return if stopped?
-        return unless processing?
-        return unless wait_thread.alive?
-        _wakeup(:event)
+        if started? && @wait_thread&.alive?
+          _wakeup(:event)
+        end
       end
 
-      def paused?
-        wait_thread && state == :paused
+      def started?
+        state == :started
       end
 
-      def processing?
-        return false if stopped?
-        return false if paused?
-        state == :processing
-      end
+      MAX_STARTUP_SECONDS = 5.0
+
+      def start
+        transition! :starting do
+          state == :pre_start or raise Error::AlreadyStarted
+        end
 
-      def setup
-        # TODO: use a Fiber instead?
-        q = ::Queue.new
         @wait_thread = Internals::ThreadPool.add do
-          _wait_for_changes(q, config)
+          _process_changes
         end
 
-        Listen::Logger.debug('Waiting for processing to start...')
-        Timeout.timeout(5) { q.pop }
-      end
+        Listen::Logger.debug("Waiting for processing to start...")
+
+        wait_for_state(:started, MAX_STARTUP_SECONDS) or raise Error::ThreadFailedToStart, "thread didn't start in #{MAX_STARTUP_SECONDS} seconds (in state: #{state.inspect})"
```

Comment:
No on `||`.

Reply:
To make this line a bit more managable, you could write:

``` ruby
unless wait_for_state(...)
  raise Error ...
end
```

Because this is very important, I'd say it's better to have explicit lines for each part.